test/experimental/test_vocab.py

```diff
@@ -0,0 +1,164 @@
+# -*- coding: utf-8 -*-
+from collections import OrderedDict
+import os
+import torch
+
+from test.common.torchtext_test_case import TorchtextTestCase
+from torchtext.experimental.vocab import (
+    Vocab
+)
+
+
+class TestVocab(TorchtextTestCase):
+    def tearDown(self):
+        super().tearDown()
+        torch._C._jit_clear_class_registry()
+        torch.jit._recursive.concrete_type_store = torch.jit._recursive.ConcreteTypeStore()
+
+    def test_has_unk(self):
+        c = OrderedDict({})
+        v = Vocab(c)
+
+        # check if unk is mapped to the first index
+        self.assertEqual(v['not_in_it'], 0)
+        self.assertEqual(v['<unk>'], 0)
+
+    def test_new_unk(self):
+        c = OrderedDict({})
+        v = Vocab(c, unk_token="<new_unk>")
+
+        # check if new_unk is mapped to the first index
+        self.assertEqual(v['<new_unk>'], 0)
+        self.assertEqual(v['not_in_it'], 0)
+
+    def test_vocab_get_item(self):
+        token_to_freq = {'a': 2, 'b': 2}
+        sorted_by_freq_tuples = sorted(token_to_freq.items(), key=lambda x: x[1], reverse=True)
+        c = OrderedDict(sorted_by_freq_tuples)
+
+        v = Vocab(c, specials=['<pad>', '<eos>'])
+        self.assertEqual(v['<unk>'], 0)
+        self.assertEqual(v['<pad>'], 1)
+        self.assertEqual(v['<eos>'], 2)
+        self.assertEqual(v['a'], 3)
+        self.assertEqual(v['b'], 4)
+
+    def test_vocab_set_item(self):
+        c = OrderedDict({'a': 2})
+
+        # add item to end
+        v = Vocab(c, specials=['<pad>', '<eos>'])
+        v['b'] = 4
+        self.assertEqual(v['<unk>'], 0)
+        self.assertEqual(v['<pad>'], 1)
+        self.assertEqual(v['<eos>'], 2)
+        self.assertEqual(v['a'], 3)
+        self.assertEqual(v['b'], 4)
+
+        # add item to middle
+        v = Vocab(c, specials=['<pad>', '<eos>'], specials_first=False)
+        v['b'] = 0
+        self.assertEqual(v['b'], 0)
+        self.assertEqual(v['a'], 1)
+        self.assertEqual(v['<unk>'], 2)
+        self.assertEqual(v['<pad>'], 3)
+        self.assertEqual(v['<eos>'], 4)
+
+    def test_vocab_add_token(self):
+        c = OrderedDict({'a': 2})
+
+        v = Vocab(c, specials=['<pad>', '<eos>'])
+        v.addToken('b')
+        self.assertEqual(len(v), 5)
+        self.assertEqual(v['b'], 4)
+
+    def test_vocab_len(self):
+        token_to_freq = {'a': 2, 'b': 2, 'c': 2}
+        sorted_by_freq_tuples = sorted(token_to_freq.items(), key=lambda x: x[1], reverse=True)
+        c = OrderedDict(sorted_by_freq_tuples)
+
+        v = Vocab(c, specials=['<pad>', '<eos>'])
+        self.assertEqual(len(v), 6)
+
+    def test_vocab_basic(self):
+        token_to_freq = {'hello': 4, 'world': 3, 'ᑌᑎIᑕOᗪᕮ_Tᕮ᙭T': 5, 'freq_too_low': 2}
+        sorted_by_freq_tuples = sorted(token_to_freq.items(), key=lambda x: x[1], reverse=True)
```

Comment:
what happen if there are tokens with same frequency?

Reply:
I believe the sorted function keeps it in the same order that it appears in. I can do some more testing on this to verify. But the whole point is that it is up to the user to provide the ordering of tokens when building the `OrderedDict`. Once the dict has been passed into the vocab class, we respect the ordering of this dict.